### ISSUE TITLE
Minimal default name mangling

### ISSUE THREAD

**Issue by TravisCardwell** [well-typed/hs-bindgen#331]
In https://github.com/well-typed/hs-bindgen/issues/312#issuecomment-2510114083 @phadej wrote:

> IMHO, the default name mangling should do as little as necessary to make valid Haskell names.
>
> Doing underscores to camelCase like translations feels a bit tricky. Perfectly we will be able to describe the mangling rules in few sentences. Users should be able to predict what the Haskell names will be. IMHO it doesn't matter what are the "idiomatic" names in Haskell.
>
> In particular, recently I started to dislike the `C` prefix. Why we have it. (`Foreign.C.Types` has it for a reason, but we don't share the same reasoning: there are no types to clash with).

In https://github.com/well-typed/hs-bindgen/issues/312#issuecomment-2510253092 @TravisCardwell replied:

> This is a question for @edsko, but I think it is just a matter of taste/preference.
>
> A few technical notes:
>
> * A prefix is required for type constructors to preserve leading underscores without translation or escaping, since type constructor names may not start with an underscore.
>
> * When preserving C names as much as possible, a prefix makes it possible to avoid converting the case of the first letter of the C name.

I discussed this with @edsko yesterday.

We currently provide two sets of name manglers.  The default (`defaultNameMangler`) creates Haskell-style type/constructor names while variables are minimally mangled.  It is safer than `haskellNameMangler`, which creates Haskell-style variable names as well.  By "safety," I mean risk of name collision.  "Safe" means relatively low risk.  In addition, the API enables developers to create their own name manglers.

@edsko is fine with making the default do as little as necessary to make valid Haskell names.  Users who want Haskell-style names can use the `haskellNameMangler`.

@phadej Do you think it is worth adding a prefix in order to avoid case conversion?  Example: C types `foo` and `Foo` may both exist.

If we do not prefix every type constructor name, we have to decide what to do about leading underscores.  Should we include a prefix only in that case?

**Comment by phadej** [comment]
> Example: C types foo and Foo may both exist.

Would they really in any reasonable library?

**Comment by phadej** [comment]
> we could generate the following, dropping the underscore and treating it like the C names are the same 

We can't. Firstly it's quite tricky to implement. Currently translation only considers the C names (and those only in quite special cases). The name mangling is an utility.

Again, the less surprising (i.e. more predictable) the pipeline is the better.

TL;DR lets keep it simple

**Comment by TravisCardwell** [comment]
Anonymous structures for named fields are not being named correctly.  Some name mangling functionality is now implemented in [`mkDefnName`][], in `HsBindgen.C.Fold.Type`.

[`mkDefnName`]: <https://github.com/well-typed/hs-bindgen/blob/d1f136cc7600a3b91f1313a767b5512194448903/hs-bindgen/src/HsBindgen/C/Fold/Type.hs#L56-L64>

If we do this, then users are not able to customize how such structures are named.  I can change the implementation to be consistent with the new name mangling defaults, but it would of course not be consistent with the `haskellNameMangler` or user-defined name manglers.

Implementing name mangling in (just) `HsBindgen.Hs.AST.Name` allows us to provide consistent name mangling across different name mangling options, as well as allows users to define their own name mangling rules.  There is already a context for this, but a difference is that the `mkDefnName` implementation uses all of the nested structure names (`Path`).  We can modify the context to provide this information.

Thoughts?

**Comment by TravisCardwell** [comment]
Unrelated issue: example [`bool.h`][] has a type `#define BOOL _Bool` that is used in a structure.  As seen in [`bool.pp.hs`][], we generate a `newtype BOOL`, but we also need to derive/generate a `Storable` instance.

[`bool.h`]: <https://github.com/well-typed/hs-bindgen/blob/d1f136cc7600a3b91f1313a767b5512194448903/hs-bindgen/examples/bool.h>
[`bool.pp.hs`]: <https://github.com/well-typed/hs-bindgen/blob/d1f136cc7600a3b91f1313a767b5512194448903/hs-bindgen/fixtures/bool.pp.hs>

**Comment by edsko** [comment]
Done in #338.
